src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -889,18 +889,24 @@ protected boolean asciiOnly(Object string,
     @CoreMethod(names = "bytes", needsBlock = true)
     public abstract static class StringBytesNode extends YieldingCoreMethodNode {
 
-        @Child private BytesNode bytesNode = BytesNode.create();
+        @Child private TruffleString.GetInternalByteArrayNode getByteArrayNode = TruffleString.GetInternalByteArrayNode
+                .create();
 
         @Specialization
         protected RubyArray bytes(Object string, Nil block,
                 @CachedLibrary(limit = "LIBSTRING_CACHE") RubyStringLibrary strings) {
-            final Rope rope = strings.getRope(string);
-            final byte[] bytes = bytesNode.execute(rope);
+            var tstring = strings.getTString(string);
+            var encoding = strings.getEncoding(string).tencoding;
+            var byteArray = getByteArrayNode.execute(tstring, encoding);
+
+            var bytes = byteArray.getArray();
+            int arrayLength = tstring.byteLength(encoding);
```

Comment:
It's convenient to use `byteArray.getLength()` in a case you already have an InternalByteArray, as that doesn't need the encoding and doesn't need to compute it again.